### DISCUSSION TITLE
Guidelines for usage in libraries, CLI apps, server apps, etc

### DISCUSSION THREAD

**Discussion by alextremblay** [KenKundert/inform#6] [Q&A]
Hello, i just recently discovered Inform, and it seems really cool!

The one thing I'm struggling to understand is the interplay between informers. 

From the documentation, it seems like any time you create an instance of the `Inform` class, it somehow overrides or temporarily shadows existing `Inform` class instances through some kind of implicit global behaviour?

Like, there's a default Inform class which is used by all informants in your code, but then if you create a new Inform class, all the informants in your code somehow know to use the settings from that informer instead of the default one. how does that work?

does the informer need to be declared in the same module / script as the informants, or can you instantiate an informer in one python module and have it affect all informants used in another python module?

Say for example, i write a python library, a package of modules meant to be imported and used by someone else's code. could i set up those modules so that they don't log or print anything by default?

And if I were to write a CLI script which imports said library, would I be able to create another informer to override that behaviour to log and print?

If i were to write a GUI app which imports this library, would I be able to configure it such that it sends messages to the system notification tray instead of stdout/stderr?

Sorry to bombard you with questions, it just seems like if all of these things are possible and documented, then Inform might well be the One Logging Library To Rule Them All 😃

**Comment by KenKundert** [comment]
There is a global variable in the inform package that holds the currently active informer. Each informer also stores the previously active informer.  You can call the disconnect method on the current informer and it will disconnect itself and activate the previous informer.  If you want to do the fancy things you described, you can use Inform with a context manager (with a _with_ statement).  So, if you wanted to call a library that also used inform, you could wrap the call in a _with_ statement, which would get you a new informer, and then you could configure that informer to disaple logging/printing and configure notifications.

You can also use the _get_informer_ and _set_informer_ functions to get or replace the currently active informer.
